hera_cal/abscal_funcs.py

```diff
@@ -1059,201 +1213,124 @@ def __eq__(self, B2):
 
-    # run abscal
-    AC = AbsCal(model, data, wgts=wgts, antpos=antpos, freqs=data_freqs, times=data_times, pols=data_pols)
-    AC.run(unravel_pol=unravel_pol, unravel_freq=unravel_freq, unravel_time=unravel_time,
-           verbose=verbose, gains2dict=True, zero_psi=zero_psi, **kwargs)
+    antpos : type=dictionary, antenna position dictionary
 
-    # smooth gains
-    if smooth:
-        AC.smooth_params()
+    flags : type=dictionary, data flags
 
-    # make gains
-    AC.make_gains()
+    broadcast_flags : type=boolean, if True, broadcast all flags across red baselines
 
-    # write to file
-    if save:
-        if calfits_fname is None:
-            calfits_fname = os.path.basename(data_file) + '.abscal.calfits'
-        AC.write_calfits(calfits_fname, overwrite=overwrite, verbose=verbose)
+    median : type=boolean, if True, take median of redundant baselines instead of mean
 
-    if output_gains:
-        return AC.gain_array
+    tol : type=float, redundant baseline tolerance threshold
 
-def abscal_arg_parser():
-    a = argparse.ArgumentParser()
-    a.add_argument("--data_files", type=str, nargs='*', help="list of miriad files of data to-be-calibrated.", required=True)
-    a.add_argument("--model_files", type=str, nargs='*', default=[], help="list of data-overlapping miriad files for visibility model.", required=True)
-    a.add_argument("--calfits_fname", type=str, default=None, help="name of output calfits file.")
-    a.add_argument("--overwrite", default=False, action='store_true', help="overwrite output calfits file if it exists.")
-    a.add_argument("--silence", default=False, action='store_true', help="silence output from abscal while running.")
-    a.add_argument("--zero_psi", default=False, action='store_true', help="set overall gain phase 'psi' to zero in linsolve equations.")
-    return a
+    mirror_red_data : type=boolean, if True, mirror average visibility across red bls
 
-
-def avg_data_across_red_bls(data, bls, antpos, flags=None, broadcast_flags=True, median=False, tol=0.5,
-                            mirror_red_data=False):
-    """
-    Given complex visibility data spanning one or more redundant
-    baseline groups, average redundant visibilities and write to file
+    Output: (red_data, red_flags, red_keys)
+    -------
     """
     # get data keys
     keys = data.keys()
 
     # get data, flags and ants
     data = copy.deepcopy(data)
+    pols = np.unique(map(lambda k: k[2], data.keys()))
     ants = np.unique(np.concatenate(keys))
     if flags is None:
-        flags = copy.deepcopy(data)
-        for k in flags.keys(): flags[k] = np.zeros_like(flags[k]).astype(np.bool)
+        flags = DataContainer(odict(map(lambda k: (k, np.zeros_like(data[k]).astype(np.bool)), data.keys())))
 
     # get redundant baselines
-    reds = compute_reds(bls, antpos, tol=tol)
+    _reds = redcal.get_reds(antpos, bl_error_tol=tol, pols=pols, low_hi=True)
+
+    # strip reds of keys not in data
```

Comment:
Low priority, but I think this can be accomplished in a line or two with list comprehension.